### PR TITLE
CoreCLR runtime tests + Mono on the x64 iOS simulator

### PR DESCRIPTION
**Pull request by imhameed**
This creates another `runtime-staging` lane, named "Build iOSSimulator x64
Release AllSubsets_Mono_RuntimeTests", that will eventually run the runtime
test suite against Mono's non-LLVM JIT on the iOS simulator on amd64 hosts.
Failing tests are added to the exclusion lists in issues.targets.

The tests aren't set to run yet, because they currently take a very long time
to execute.

`AppleAppBuilder` no longer requires a `MainLibraryFileName`. If omitted, one
must be supplied when the app bundle is launched via an environment variable
named `MONO_APPLE_APP_ENTRY_POINT_LIB_NAME`. The generated apps also accept
another environment variable named `MONO_APPLE_APP_ASSEMBLY_LOAD_PREFIX`, which
is a hack used to allow assembly lookup to proceed in a nested app-relative
subdirectory before falling back to the root of the app bundle. This is
necessary because app bundles contain multiple individual test assemblies, and
these assemblies sometimes have dependencies with names that collide with the
dependencies of other test assemblies inside the bundle.

### CODE REVIEW

**Comment by imhameed** [comment]
Current status: can build apps containing entire runtime test categories. Can also read stdout from instrumented native app launcher code from `xcrun simctl launch --console-pty`. Need to figure out why the tests don't seem to actually run (despite the runtime superficially launching correctly).

**Comment by imhameed** [comment]
Haven't spent much time on this this week; have mostly been occupied with arm64 simd.

**Comment by imhameed** [comment]
I got the tests running in the simulator. I also tried getting these running on-device, but that will require a bit more work, and won't be a part of this PR.

**Comment by SamMonoRT** [comment]
No progress made in last few days, will pick up this work next week.

**Comment by SamMonoRT** [comment]
@imhameed working on this item this week

**Comment by imhameed** [comment]
baseservices_mono.app/runningmono.dll passed locally:

```
XHarness command issued: apple run --app=baseservices_mono.app --output-directory=out --targets=ios-simulator-64 --mlaunch=/Library/Frameworks/Xamarin.iOS.framework/Versions/14.10.0.4/bin/mlaunch --set-env=MONO_APPLE_APP_ENTRY_POINT_LIB_NAME=runningmono.dll --expected-exit-code=100
info: Preparing run for ios-simulator-64
info: Getting app bundle information..
info: Starting application 'baseservices_mono' on ios-simulator-64
info: App run ended with 100
info: Application has finished with exit code: 100 (as expected)
XHarness exit code: 0
```

**Comment by imhameed** [comment]
And xharness now supports app installation and app execution as separate steps (but app uninstalling doesn't yet work in the simulator), after an update to mlaunch: https://github.com/dotnet/xharness/issues/493

```
xharness.sh apple install --app=baseservices_mono.app --output-directory=out --targets=ios-simulator-64 --mlaunch=/Users/imhameed/ms/xharness/artifacts/tmp/Debug/macios-binaries/mlaunch/bin/mlaunch           
XHarness command issued: apple install --app=baseservices_mono.app --output-directory=out --targets=ios-simulator-64 --mlaunch=/Users/imhameed/ms/xharness/artifacts/tmp/Debug/macios-binaries/mlaunch/bin/mlaunch
info: Preparing run for ios-simulator-64
info: Getting app bundle information from '/Users/imhameed/ms/dnr-9.x-ios/runtime/artifacts/tests/coreclr/obj/iOSSimulator.x64.Release/tests/run/iOSApps/baseservices_mono/AppBundle/baseservices_mono/Release-iphonesimulator/baseservices_mono.app'
info: Looking for available ios-simulator-64 simulators..
info: Found simulator device 'iPhone X (iOS 14.4) - created by XHarness'
info: Installing application 'baseservices_mono' on 'iPhone X (iOS 14.4) - created by XHarness'
info: Application 'baseservices_mono' was installed successfully on 'iPhone X (iOS 14.4) - created by XHarness'
XHarness exit code: 0
```
```
xharness.sh apple just-run --app=baseservices_mono.app --output-directory=out --targets=ios-simulator-64 --mlaunch=/Users/imhameed/ms/xharness/artifacts/tmp/Debug/macios-binaries/mlaunch/bin/mlaunch --set-env=MONO_APPLE_APP_ENTRY_POINT_LIB_NAME=runningmono.dll --expected-exit-code=100
XHarness command issued: apple just-run --app=baseservices_mono.app --output-directory=out --targets=ios-simulator-64 --mlaunch=/Users/imhameed/ms/xharness/artifacts/tmp/Debug/macios-binaries/mlaunch/bin/mlaunch --set-env=MONO_APPLE_APP_ENTRY_POINT_LIB_NAME=runningmono.dll --expected-exit-code=100
info: Preparing run for ios-simulator-64
info: Getting app bundle information from '/Users/imhameed/ms/dnr-9.x-ios/runtime/artifacts/tests/coreclr/obj/iOSSimulator.x64.Release/tests/run/iOSApps/baseservices_mono/AppBundle/baseservices_mono/Release-iphonesimulator/baseservices_mono.app'
info: Looking for available ios-simulator-64 simulators..
info: Found simulator device 'iPhone X (iOS 14.4) - created by XHarness'
info: Starting application 'baseservices_mono' on 'iPhone X (iOS 14.4) - created by XHarness'
info: App run ended with 100
info: Application has finished with exit code: 100 (as expected)
XHarness exit code: 0
```

**Comment by premun** [comment]
You can disregard that apps don't get uninstalled. It's been like this for years for Xamarin and we also test the whole BCL like that in dotnet/runtime. I don't think the apps get installed in the same way as for real devices and it doesn't cause issues.

If we implement this in the future, it will just start working for you out of the box (if you decide to start calling uninstall now).

File: eng/pipelines/runtime.yml
Comment (SamMonoRT) [inline eng/pipelines/runtime.yml:583]: This is in draft, but just a heads-up - New lanes should be added to the runtime staging yaml

**Comment by SamMonoRT** [comment]
@imhameed - I believe you need an entry in ...eng/pipelines/coreclr/templates/helix-queues-setup.yml file to actually kick off the runtime tests. Looking at the iOS simulator results, doesn't seem like it ran the tests.

File: src/tests/Common/CLRTest.Execute.Bash.targets
Comment (premun) [inline src/tests/Common/CLRTest.Execute.Bash.targets:390]: The problem with running iOS Simulator workloads in Helix is that the way the Helix client is running doesn't work with the Simulator as there's no user GUI session. We work around this by executing the whole payload using `sudo launchctl asuser`. There are couple of more tricks like this (re-launching the `Simulator.app`, detection of Simulator hangs, cleanup..).

Since there are so many caveats and to improve stability, I have tried to capture all this in Helix SDK, specifically in these scripts:
- https://github.com/dotnet/arcade/blob/main/src/Microsoft.DotNet.Helix/Sdk/tools/xharness-runner/xharness-helix-job.android.sh
- https://github.com/dotnet/arcade/blob/main/src/Microsoft.DotNet.Helix/Sdk/tools/xharness-runner/xharness-helix-job.apple.sh
- https://github.com/dotnet/arcade/blob/main/src/Microsoft.DotNet.Helix/Sdk/tools/xharness-runner/xharness-helix-job.apple.sh

Currently, when you define your own command and only ask for the XHarness CLI to be pre-installed, we don't then run your script as part of this wrapper and this will have to be fixed (we have an issue for that that you've seen already https://github.com/dotnet/xharness/issues/590)

Comment (premun) [reply src/tests/Common/CLRTest.Execute.Bash.targets:390]: With what you try here - you will get timeouts on the simulators (the process spawns but we can't talk to it)

Comment (premun) [reply src/tests/Common/CLRTest.Execute.Bash.targets:390]: I just got this work prioritized because I see it will be a blocker for you. I estimate around 4 days of work so it would be ready next week. We will then have to come back here and check how you turn your MSBuild items into `HelixWorkItem` so that we can plug it in with what I prepare.

**Comment by premun** [comment]
@imhameed can you help me to identify which build and which build leg is this PR creating?

I would like to see the binlog from the build to see how you create the Helix job

**Comment by SamMonoRT** [comment]
> @imhameed can you help me to identify which build and which build leg is this PR creating?
> 
> I would like to see the binlog from the build to see how you create the Helix job

It is creating this runtime staging lane : https://github.com/dotnet/runtime/pull/43954/checks?check_run_id=2510681839

**Comment by premun** [comment]
I checked how you run Helix and I think the best way to approach this is that in **helixpublishwitharcade.proj**, you will need to create `<XHarnessHelixWorkItem>` instead of `<HelixWorkItem>`. I will then make it include the script inside of the Helix wrapper scripts that will prepare the simulators/devices and clean up after.

Do you share this `.proj` file with others who have non-XHarness workloads? Or you're the only users?

**Comment by imhameed** [comment]
Weird: the set of tests that failed in the most recent run is not exactly the same as the set of tests that failed in the previous run. Some of the failures are reproducible locally but many are not.

**Comment by imhameed** [comment]
Related: https://github.com/dotnet/xharness/issues/596

**Comment by imhameed** [comment]
The new lane went green: https://dev.azure.com/dnceng/public/_build/results?buildId=1143523&view=logs&j=4cb2c771-3167-5285-9e21-64bc263d1f6f&t=d48eef33-240c-5153-63be-598e59e13d53

`dci-mac-build-143` took a very long time to run `JIT.Methodical.eE`.

**Comment by premun** [comment]
@imhameed can you point me to a log file where you can see the xharness logs? How we call XHarness and which commands?

**Comment by imhameed** [comment]
xharness is called by this script: https://github.com/dotnet/runtime/pull/43954/files#diff-3e5a329159ca5b2268e62be8a0d776b6092681e9b241210cb4d57e3454816abcR373-R415

Only `xharness apple run` is used right now. (Android is similar.) Here's the most recent run: https://helix.dot.net/api/jobs/01a12fa7-0660-4bdb-bd7d-d956f19999ee/workitems?api-version=2019-06-17

An example log for `JIT.Methodical.eE`:
https://helix.dot.net/api/jobs/01a12fa7-0660-4bdb-bd7d-d956f19999ee/workitems/JIT.Methodical.eE?api-version=2019-06-17
https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-43954-merge-01a12fa706604bdbbd/JIT.Methodical.eE/console.2a816a29.log?sv=2019-07-07&se=2021-06-07T13%3A32%3A04Z&sr=c&sp=rl&sig=Io09wf8ykr%2BTWHE2jmHNkqPR7k3Obpvzo%2BL%2BaCpyijE%3D

I haven't yet replaced the existing setup with a `CustomCommand` setup that strips away the "XUnitWrapper" and script-invoked-by-XUnitWrapper layers of indirection. I am increasingly coming to believe that it'd make more sense to leave the execution model of this as-is, and then clean up both Android and iOS simultaneously after this has been merged.

File: src/tests/issues.targets
Comment (SamMonoRT) [inline src/tests/issues.targets:3413]: For this and all below, it is recommended to add GH issue links here. If you haven't yet created these, I would suggest updating this file as a subsequent PR once this is merged.

File: src/tests/Common/CLRTest.Execute.Bash.targets
Comment (SamMonoRT) [inline src/tests/Common/CLRTest.Execute.Bash.targets:369]: nit : Exit code

File: eng/pipelines/coreclr/templates/helix-queues-setup.yml
Comment (SamMonoRT) [inline eng/pipelines/coreclr/templates/helix-queues-setup.yml:31]: for my knowledge, will this be used for x86 too ?

File: src/tests/issues.targets
Comment (imhameed) [reply src/tests/issues.targets:3413]: That's the plan.

**Comment by SamMonoRT** [comment]
@imhameed - Great work. Please outline any follow up work here based on discussions with Premek. Also, please indicate how many tests are actually running, and how many are excluded at the moment.

File: src/tests/issues.targets
Comment (SamMonoRT) [reply src/tests/issues.targets:3413]: It might make sense to create issues prior to adding new iOS/tvOS device lanes to easily build over common failures.

File: eng/pipelines/coreclr/templates/helix-queues-setup.yml
Comment (imhameed) [reply eng/pipelines/coreclr/templates/helix-queues-setup.yml:31]: There are no plans, as far as I know, for an iOS Simulator x86 lane. This was duplicated from `pipelines/libraries/helix-queues-setup.yml`. I can remove the definitions platforms we definitely are not testing as part of https://github.com/dotnet/runtime/issues/33381.

**Comment by SamMonoRT** [comment]
LGTM - Approving, but need another reviewer to approve prior to merging.

**Review by SamMonoRT — APPROVED** [review]

**Comment by imhameed** [comment]
The most recent build had a 4-hour timeout expire on the new lane: https://dev.azure.com/dnceng/public/_build/results?buildId=1144258&view=logs&j=4cb2c771-3167-5285-9e21-64bc263d1f6f&t=4cb2c771-3167-5285-9e21-64bc263d1f6f

In this run, the long-running job was `JIT.Methodical`. It ran on `dci-mac-build-146`. All other lanes took much less time to run.

The long-running job from the last test run, `JIT.Methodical.eE`, took [2h44m37s to complete](https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-43954-merge-01a12fa706604bdbbd/JIT.Methodical.eE/console.2a816a29.log?sv=2019-07-07&se=2021-06-07T13%3A32%3A04Z&sr=c&sp=rl&sig=Io09wf8ykr%2BTWHE2jmHNkqPR7k3Obpvzo%2BL%2BaCpyijE%3D) on `dci-mac-build-143`. On this run, it took [37m52s to complete](https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-43954-merge-5b695068d2a84a9c84/JIT.Methodical.eE/console.1c7a0a2b.log?sv=2019-07-07&se=2021-06-07T20%3A27%3A38Z&sr=c&sp=rl&sig=4Eo5klO2n2mPVOTelh3%2Baj8vdF%2FQpEPWWgHJ7zb%2BSV4%3D) on `dci-macpro-07`.

**Comment by premun** [comment]
The queue has MacMinis and MacPros. I imagine MacPros will be way faster, yeah.

**Comment by premun** [comment]
@SamMonoRT @imhameed we should definitely not check this in without using the Helix SDK properly (aka `CustomCommands`). I worked with Imran to do the bare minimum to be able to run the simulators in this PR (such as the `launchctl asuser`) but it was never my intention to get this checked in without switching onto the proper usage. The reason is that we do quite a lot of work in the Helix SDK to make sure these workloads run well on Helix and the only way for us to do this by running some code before and after the actual workload.

Not using the Helix SDK properly results in:
- Improper setup and cleanup of the machines. For example it happens about once a week that a Simulator freezes and we resolve that by rebooting the machine and requesting a retry on a different machine. Other than that we kill and prepare the Simulator process so that you don't experience failures.
- Inability to sign apps for device targets (I don't think you care about this at this point, but you might in the future)
- Uploading of `testResults.xml` to AzDO

It's not that you cannot copy the SDK scripts and do similar things here, part of that already happened in this PR just to make it even work. This is more about us needing to change things and deploy them everywhere quite often. For example there is  upcoming work around telemetry where we need to report well for XHarness scenarios which failures are infrastructural and which are caused by the client.

Can we still switch over the to the proper usage before checking in?

File: src/tests/Directory.Build.targets
Comment (fanyang-mono) [inline src/tests/Directory.Build.targets:174]: Add apple to the comment as well.

**Review by fanyang-mono — APPROVED** [review]
Great work!

**Comment by SamMonoRT** [comment]
> @SamMonoRT @imhameed we should definitely not check this in without using the Helix SDK properly (aka `CustomCommands`). I worked with Imran to do the bare minimum to be able to run the simulators in this PR (such as the `launchctl asuser`) but it was never my intention to get this checked in without switching onto the proper usage. The reason is that we do quite a lot of work in the Helix SDK to make sure these workloads run well on Helix and the only way for us to do this by running some code before and after the actual workload.
> 
> Not using the Helix SDK properly results in:
> 
> * Improper setup and cleanup of the machines. For example it happens about once a week that a Simulator freezes and we resolve that by rebooting the machine and requesting a retry on a different machine. Other than that we kill and prepare the Simulator process so that you don't experience failures.
> * Inability to sign apps for device targets (I don't think you care about this at this point, but you might in the future)
> * Uploading of `testResults.xml` to AzDO
> 
> It's not that you cannot copy the SDK scripts and do similar things here, part of that already happened in this PR just to make it even work. This is more about us needing to change things and deploy them everywhere quite often. For example there is upcoming work around telemetry where we need to report well for XHarness scenarios which failures are infrastructural and which are caused by the client.
> 
> Can we still switch over the to the proper usage before checking in?

@imhameed  - we should understand and make the relevant changes. This will help with the device tests which are in the pipeline.

File: src/tests/Common/CLRTest.Execute.Bash.targets
Comment (premun) [inline src/tests/Common/CLRTest.Execute.Bash.targets:402]: Question - the overall Helix job exits with this if at least once you get non-0?

**Comment by premun** [comment]
@SamMonoRT @imhameed I did some digging today into the Helix job this new build leg creates and it adds ~16 hours of build time to the `OSX.1015.Amd64.Open` queue (approx 6 times more than the BCL test suite does).

We should definitely not merge this yet because:
- Running this in every `runtime-staging` build will overload the queue which doesn't have the capacity yet - currently, BCL tests already take a good chunk and on busy days it is so-so
- One of the work items here takes almost 4 hours so it needs to be partitioned
- We should do the proper Helix SDK integration

@ilyas1974 @tkapin we should think about how we're going to support these new runtime tests in Helix as we might need way more capacity.
We can figure out some temporary solution too like utilizing one of the Apple device queues which currently get zero traffic.

**Comment by SamMonoRT** [comment]
> @SamMonoRT @imhameed I did some digging today into the Helix job this new build leg creates and it adds ~16 hours of build time to the `OSX.1015.Amd64.Open` queue (approx 6 times more than the BCL test suite does).
> 
> We should definitely not merge this yet because:
> 
> * Running this in every `runtime-staging` build will overload the queue which doesn't have the capacity yet - currently, BCL tests already take a good chunk and on busy days it is so-so
> * One of the work items here takes almost 4 hours so it needs to be partitioned
> * We should do the proper Helix SDK integration
> 
> @ilyas1974 @tkapin we should think about how we're going to support these new runtime tests in Helix as we might need way more capacity.
> We can figure out some temporary solution too like utilizing one of the Apple device queues which currently get zero traffic.

Yes, lets iron this out prior to merging : 
1. @imhameed - please update the PR with the customCommands usage and have a green passing lane.
2.  @premun @tkapin : We should discuss a plan to enable these runs. Also note, Imran will be working on trying to add the iOS device lanes soon, also adding an interpreter lane for the simulator as well. We need to make sure there is enough hardware to handle the load, and if there are any tips he should consider while he scripts those up. 
3. cc @marek-safar @steveisok as fyi

File: src/tests/Common/CLRTest.Execute.Bash.targets
Comment (akoeplinger) [inline src/tests/Common/CLRTest.Execute.Bash.targets:394]: For android we introduced separate install/run/uninstall commands so we don't install/uninstall the app a thousand times during the test run which helped improve the time a lot, we probably need the same here: https://github.com/dotnet/xharness/issues/397

Comment (premun) [reply src/tests/Common/CLRTest.Execute.Bash.targets:394]: We have the same commands available in iOS now: https://github.com/dotnet/xharness/blob/98970876f4d21e6da5198a5c41de51ab1101089f/tests/integration-tests/Apple/Simulator.Commands.Tests.proj#L24-L41

https://github.com/dotnet/xharness/issues/493

Comment (akoeplinger) [reply src/tests/Common/CLRTest.Execute.Bash.targets:394]: hmm it's a pity that the `android run` command does run-without-install but the `apple run` command does install+run+uninstall and you need `apple just-test`. can we look at coming up with some unified naming in xharness?

Comment (premun) [reply src/tests/Common/CLRTest.Execute.Bash.targets:394]: Yeah, there's a discrepancy because Apple has two different runtime modes - when it just runs the app and when it does the whole TestRunner listening and whatnot.

I think it would make sense to rename the Android command to `just-test` and have `test` and `just-test` in Android and Apple and then the `run` and `just-run` variants in Apple only?

Comment (imhameed) [reply src/tests/Common/CLRTest.Execute.Bash.targets:394]: > For android we introduced separate install/run/uninstall commands so we don't install/uninstall the app a thousand times during the test run which helped improve the time a lot, we probably need the same here: [dotnet/xharness#397](https://github.com/dotnet/xharness/issues/397)

I don't think `xharness android install` is being used at all right now by the runtime tests.

**Comment by premun** [comment]
@SamMonoRT @imhameed we spoke with @ilyas1974 and @tkapin about our options and we think that we first need you to optimize the workloads (for exmaple by not re-installing the applications over and over) to see what kind of build time we're dealing with.

After that we have several options how to proceed such as utilizing other OSX queues (such as the 11.00 one or the AppleTV one) which don't have much traffic at the moment and we will help you target the queues so that we don't overload any single one queue.

**Comment by fanyang-mono** [comment]
> @SamMonoRT @imhameed we spoke with @ilyas1974 and @tkapin about our options and we think that we first need you to optimize the workloads (for exmaple by not re-installing the applications over and over) to see what kind of build time we're dealing with.
> 
> After that we have several options how to proceed such as utilizing other OSX queues (such as the 11.00 one or the AppleTV one) which don't have much traffic at the moment and we will help you target the queues so that we don't overload any single one queue.

Just a note here, our team (Sam's team) is looking into enable runtime tests on AppleTV as well. Utilizing AppleTV queue could only be a very temporary solution here. Here is the overall plan of Mono testing. https://github.com/dotnet/runtime/issues/33381 This should help with queue planing.

In terms of the utilizing `android install/run/uninstall`,  I am expecting to work on that with the next week or so.

**Comment by tkapin** [comment]
Perfect. We really need to base any utilization estimates on efficient usage of the resources and using the install/test/uninstall commands is supposed to improve the efficiency significantly. Also, it's important to not harm any of the other teams that are using the shared resources.

File: src/tests/Common/CLRTest.Execute.Bash.targets
Comment (fanyang-mono) [reply src/tests/Common/CLRTest.Execute.Bash.targets:392]: When https://github.com/dotnet/runtime/pull/54543 gets merged, you should be able to use test only(no more app installation needed here) apple xharness command here.

**Comment by fanyang-mono** [comment]
Tests failed to run and apps failed to uninstall. Currently blocked by https://github.com/dotnet/core-eng/issues/13678

**Comment by imhameed** [comment]
@fanyang-mono @premun I'm going to merge this PR. I've disabled sending the tests to Helix, so this PR won't result in increased use of the OSX.1015.Amd64.Open pool. Followup work (like using CustomCommands, or completing the transition to `xharness apple just-run`, or reducing the end-to-end runtime of a fully-enabled iOSSimulator runtime test lane) will build on the work done in this PR.

**Review by akoeplinger — COMMENTED** [review]
LGTM

**Pull request merged**
